File: OrganizationService/src/main/java/com/slavacom/organizationservice/config/LoggingClientHttpRequestInterceptor.java
package com.slavacom.organizationservice.config;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Component
public class LoggingClientHttpRequestInterceptor implements ClientHttpRequestInterceptor {

    private static final Logger logger = LoggerFactory.getLogger(LoggingClientHttpRequestInterceptor.class);

    @Override
    public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution)
            throws IOException {
        long startTime = System.currentTimeMillis();

        String method = request.getMethod().name();
        String uri = request.getURI().toString();

        logger.debug("{} {} - request started", method, uri);

        try {
            ClientHttpResponse response = execution.execute(request, body);
            long duration = System.currentTimeMillis() - startTime;

            logger.info("{} {} completed in {}ms with status {}",
                    method,
                    uri,
                    duration,
                    response.getStatusCode().value());

            return response;
        } catch (IOException e) {
            long duration = System.currentTimeMillis() - startTime;
            logger.error("{} {} failed in {}ms with error: {}",
                    method,
                    uri,
                    duration,
                    e.getMessage(),
                    e);
            throw e;
        }
    }
}

File: NotificationService/src/main/kotlin/com/slavacom/notificationservice/config/LoggingClientHttpRequestInterceptor.kt
package com.slavacom.notificationservice.config

import mu.KotlinLogging
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import org.springframework.stereotype.Component
import java.io.IOException

private val logger = KotlinLogging.logger {}

@Component
class LoggingClientHttpRequestInterceptor : ClientHttpRequestInterceptor {

    override fun intercept(
        request: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution
    ): ClientHttpResponse {
        val startTime = System.currentTimeMillis()
        val method = request.method.name
        val uri = request.uri.toString()

        logger.debug { "$method $uri - request started" }

        return try {
            val response = execution.execute(request, body)
            val duration = System.currentTimeMillis() - startTime

            logger.info { "$method $uri completed in $duration ms with status ${response.statusCode.value()}" }
            response
        } catch (e: IOException) {
            val duration = System.currentTimeMillis() - startTime
            logger.error(e) { "$method $uri failed in $duration ms with error: ${e.message}" }
            throw e
        }
    }
}
